lint/param-duplication_checker.go

```diff
@@ -0,0 +1,70 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+// ParamDuplicationChecker detects functions where parameters declaration
+// could be simplified by combine arguments with same type
+//
+// Example: func f(a int, b int) could be simplified as func f(a, b int)
+type ParamDuplicationChecker struct {
+	ctx *Context
+
+	warnings []Warning
+}
+
+// NewParamDuplicationChecker returns initialized ParamDuplicationChecker
+func NewParamDuplicationChecker(ctx *Context) *ParamDuplicationChecker {
+	return &ParamDuplicationChecker{
+		ctx: ctx,
+	}
+}
+
+// Check runs parameter duplication checks for f.
+//
+// Features
+//
+// Detects if function parameters could be combined by type
+// and suggest the way to do it.
+func (c *ParamDuplicationChecker) Check(f *ast.File) []Warning {
+	c.warnings = c.warnings[:0]
+	for _, decl := range collectFuncDecls(f) {
+		c.checkParamDuplication(decl)
+	}
+	return c.warnings
+}
+
+// TODO(fexolm) don't create multiple warnings on the same function
+// TODO(fexolm) create warning in other function
+func (c *ParamDuplicationChecker) checkParamDuplication(decl *ast.FuncDecl) {
+	params := decl.Type.Params.List
+	if len(params) < 2 {
+		return
+	}
+	for i, p := range params[1:] {
+		if astcmp.EqualExpr(p.Type, params[i].Type) {
+			var winfo string
```

Comment:
We'll find some way to make message construction easier later.
OK for now.
Thought you could move whole block into it's own method, like `warn`.

Reply:
I have TODO for this :)

Reply:
Do you have particular reasons for this?

I do believe that TODO are only good for things that you are not going to do in a few days or hours. If it's a matter of moving code into a method, it's better to do it right away, at least that's my preference.

Reply:
Done